app/utils/sparse-fieldsets.ts

```diff
@@ -0,0 +1,70 @@
+import { camelize, underscore } from '@ember/string';
+import ModelRegistry from 'ember-data/types/registries/model';
+import { pluralize, singularize } from 'ember-inflector';
+import { Links } from 'jsonapi-typescript';
+
+import { Resource } from 'osf-api';
+
+// Automatic type safety for attributes and relationships appears prohibitively expensive.
+// Instead, use `unknown` to force explicit types.
+export type SparseModel = {
+    id: string | number;
+    modelName: string;
+    links?: Links;
+} & Record<string, unknown>;
+
+export type SparseFieldset = {
+    [K in keyof ModelRegistry]?: string[]
+};
+
+function mapObject<T, R>(
+    obj: T,
+    mapKey: (k: string) => string,
+    mapValue: (v: T[keyof T]) => R,
+): Record<string, R> {
+    return Object.entries(obj).reduce(
+        (acc, [k, v]) => ({
+            ...acc,
+            [mapKey(k)]: mapValue(v),
+        }),
+        {} as Record<string, R>,
+    );
+}
+
+function extractEmbeds(resource: Resource): Record<string, SparseModel | SparseModel[]> {
+    return mapObject(
+        resource.embeds || {},
+        camelize,
+        /* eslint-disable no-use-before-define */
+        v => (v.data instanceof Array ?
```

Comment:
Why the eslint disable here?

Reply:
`extractEmbeds` and `parseSparseResource` are mutually recursive, so eslint would complain no matter their order. This seems like a natural case for mutual recursion, since embeds are full resource documents in their own right.

I should definitely explain myself when I disable linting, though.

Reply:
Yea, not sure about this one, its incredibly hard to reason about. Worried that it will be brittle (in terms of updating and changing). I'm not even sure if you should have issues from mutual recursion - or, even, mutual recursion at all. Is this just mapping data from one for to another?

Reply:
Yeah, it's flattening the API response into a simpler form, with camelCase keys to keep our code consistent. The recursion was to handle embedded resources, which have the exact same schema as the top-level resource (and may have their own embedded resources).

I moved things around so it's just normal recursion, and came out with something much more readable, I think.